### PR TITLE
Auth: cache user passwords

### PR DESCRIPTION
**Pull request by nirbhayc**
Signed-off-by: Nirbhay Choubey <nchoubey@bloomberg.net>

### CODE REVIEW

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/fe00fe5c/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: Success &#10003;.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/fe00fe5c/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/fe00fe5c/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2686'>Resubmit</a></p>

File: bdb/llmeta.c [hunk 1]
```diff
@@ -9690,6 +9693,51 @@ int bdb_user_exists(tran_type *tran, char *user)
     return rc;
 }
 
+int gbl_max_password_cache_size = 100;
+lrucache *password_cache; // saved password hashes
```
Comment (riverszhang89) [inline bdb/llmeta.c:9697]: static

File: bdb/llmeta.c [hunk 2]
```diff
@@ -9690,6 +9693,51 @@ int bdb_user_exists(tran_type *tran, char *user)
...
+    uint8_t *key = (uint8_t *)p;
+    for (int i = 0; i < PASSWD_HASH_SZ; ++ i) {
+        h = ((h % 8388013) << 8) + (key[i]);
+    }
+    return h;
+}
+
+static int password_cmp(const void *key1, const void *key2, int len)
+{
+    return memcmp(key1, key2, PASSWD_HASH_SZ);
```
Comment (riverszhang89) [inline bdb/llmeta.c:9718]: CRYPTO_memcmp

**Review by riverszhang89 — CHANGES_REQUESTED** [review]

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

**Comment by riverszhang89** [comment]
Looks great. Would you want to squash the commits so that I can approve?

**Review by mponomar — COMMENTED** [review]
I don't see an lrucache_release?  It'd be needed for things to get removed from the list.

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3e1ef50e/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>5/400 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>lostwrite [setup failure]<br>sc_timepart_logicalsc_generated<br>simple_timepart_reptimeout_generated<br>simple_timepart<br>sc_timepart</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3e1ef50e/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/3e1ef50e/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2686'>Resubmit</a></p>

**Comment by nirbhayc** [comment]
> Looks great. Would you want to squash the commits so that I can approve?

Done

**Comment by nirbhayc** [comment]
> I don't see an lrucache_release? It'd be needed for things to get removed from the list.

I didn't know that. Added in the latest commit. Thank you for pointing.

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e32d1509/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>3/401 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>simple_timepart_reptimeout_generated<br>sc_timepart_logicalsc_generated<br>simple_timepart</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e32d1509/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e32d1509/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2686'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e32d1509/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e32d1509/robomark.txt'>Error &#x26A0; </a>.<br>Regression testing: <b>2/405 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>op_makerecord<br>tunables</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e32d1509/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/e32d1509/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2686'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/938148f1/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>1/405 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>pmux</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/938148f1/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/938148f1/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2686'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/formatting-a-pr.html'>Use clang-format</a> or <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/938148f1/format.diff'>view diff</a> &#x26A0;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>1/405 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>op_makerecord</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/938148f1/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/938148f1/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2686'>Resubmit</a></p>

**Review by riverszhang89 — APPROVED** [review]

**Comment by nirbhayc** [comment]
@riverszhang89 Thanks!

**Pull request merged**